### PR TITLE
[x86/Linux] Fix IL Stub for Shared Generic

### PR DESCRIPTION
**Pull request by seanshpark**
Fix parameter order in UnboxingILStub for Shared Generic

### CODE REVIEW

**Comment by seanshpark** [comment]
Similar patch to #8848

**Comment by seanshpark** [comment]
Fixes #9480

**Comment by seanshpark** [comment]
CC @parjong, @wateret , @YongseopKim

**Comment by seanshpark** [comment]
@jkotas , @janvorli , PTAL

File: src/vm/prestub.cpp [hunk 1]
```diff
@@ -731,31 +731,52 @@ Stub * CreateUnboxingILStubForSharedGenericValueTypeMethods(MethodDesc* pTargetM
...
 
-    // 2.2 Push the hidden context param 
-    // The context is going to be captured from the thisptr
-    pCode->EmitLoadThis();
-    pCode->EmitLDFLDA(tokPinningHelper);
-    pCode->EmitLDC(Object::GetOffsetOfFirstField());
-    pCode->EmitSUB();
-    pCode->EmitLDIND_I();
+#if defined(_TARGET_X86_)
+    if (numArgs < NUM_ARGUMENT_REGISTERS)
```
Comment (jkotas) [inline src/vm/prestub.cpp]: I do not think you should be checking the number of arguments here. The generic context should be always the last argument on x86, no exceptions.

E.g. see https://github.com/dotnet/coreclr/blob/4822563366cec72e3e269e56a14a15bb238ba8b2/src/vm/callingconvention.h#L1515

Comment (jkotas) [inline src/vm/prestub.cpp]: I see that a similar fix was done in CreateInstantiatingILStub some time ago as well. It should be fixed there as well to not check NUM_ARGUMENT_REGISTERS - if you agree that it is the right fix.

Comment (seanshpark) [inline src/vm/prestub.cpp]: Then I think this can be a little bit simple. As `CreateInstantiatingILStub` change was from @janvorli , is it OK?

**Comment by seanshpark** [comment]
If don't have to check num of arguments it could be something like this.
```
diff --git a/src/vm/prestub.cpp b/src/vm/prestub.cpp
index 723e6b6..7862101 100644
--- a/src/vm/prestub.cpp
+++ b/src/vm/prestub.cpp
@@ -738,7 +738,15 @@ Stub * CreateUnboxingILStubForSharedGenericValueTypeMethods(MethodDesc* pTargetM
     pCode->EmitLoadThis();
     pCode->EmitLDFLDA(tokPinningHelper);
 
-    // 2.2 Push the hidden context param 
+#if defined(_TARGET_X86_)
+    // 2.2 Push the rest of the arguments
+    for (unsigned i = 0; i < msig.NumFixedArgs();i++)
+    {
+        pCode->EmitLDARG(i);
+    }
+#endif
+
+    // 2.3 Push the hidden context param 
     // The context is going to be captured from the thisptr
     pCode->EmitLoadThis();
     pCode->EmitLDFLDA(tokPinningHelper);
@@ -746,11 +754,13 @@ Stub * CreateUnboxingILStubForSharedGenericValueTypeMethods(MethodDesc* pTargetM
     pCode->EmitSUB();
     pCode->EmitLDIND_I();
 
-    // 2.3 Push the rest of the arguments
+#if !defined(_TARGET_X86_)
+    // 2.4 Push the rest of the arguments
     for (unsigned i = 0; i < msig.NumFixedArgs();i++)
     {
         pCode->EmitLDARG(i);
     }
+#endif
 
     // 2.4 Push the target address
     pCode->EmitLDC((TADDR)pTargetMD->GetMultiCallableAddrOfCode(CORINFO_ACCESS_ANY));
```

**Comment by jkotas** [comment]
Right - this looks better to me.

File: src/vm/prestub.cpp (hunk 1 above)
Comment (janvorli) [inline src/vm/prestub.cpp]: @jkotas so you are saying we should make it different from Windows x86?

**Comment by seanshpark** [comment]
@jkotas, @janvoli, PTAL once more, thanks!

**Comment by janvorli** [comment]
I am confused by how this could work without corresponding change in the JIT. As I have recently added to the ABI doc (based on looking into where the JITted code expects the arguments while you were fixing the  CreateInstantiatingILStub):
> On x86, if all arguments of the function including "this" pointer fit into argument registers (ECX and EDX) and we still have argument registers available, we store the hidden argument in the next available argument register. Otherwise it is passed as the last stack argument.

@jkotas did you mean to change the JIT accordingly or am I missing something?

**Comment by jkotas** [comment]
I do not think that the ABI doc describes what we do on Windows accurately. Take this example:

```
struct BigStruct
{
    int a, b, c, d, e, f;
}

class Program
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static void MyG<T>(BigStruct s1, BigStruct s2)
    {
    }

    static void Main(string[] args)
    {
        MyG<string>(new BigStruct(), new BigStruct());
    }
}
```

The callsite in the main method looks like this:

```
... push s1 and s2 on the stack ...
mov         ecx,49A0F90h ... hidden argument
call        dword ptr ds:[49A0F64h]
```

As you can see the hidden argument is passed in register because of we have still register available. But according to the ABI doc, it should have been passed on the stack because of all arguments of the function do not fit into argument registers.

Instead, it should simply say that the hidden argument is passed in as last argument.

**Comment by jkotas** [comment]
@seanshpark LGTM. Could you please also fix `CreateInstantiatingILStub` to match, and also fix the incorrect description in the ABI doc ?

**Comment by janvorli** [comment]
@jkotas I can see now my description was a bit confusing, I really meant  "if all arguments of the function **that can be passed in registers** including "this" pointer fit into argument registers". And I haven't realized it is in fact equivalent to saying that the hidden argument is passed as the last argument and that the rules of passing parameters in registers / on stack take care of the actual location.

**Comment by jkotas** [comment]
Ok. I have opened https://github.com/dotnet/coreclr/issues/9577 to make the same fix in CreateInstantiatingILStub as well.

**Comment by seanshpark** [comment]
@jkotas , @janvorli , Thank you both. I'll make a PR to fix CreateInstantiatingILStub like this.

**Pull request merged**